src/auth/flows.rs

```diff
@@ -887,3 +896,155 @@ pub async fn auth_callback(
+    req: HttpRequest,
+    pool: Data<PgPool>,
+    redis: Data<deadpool_redis::Pool>,
+    new_account: web::Json<NewAccount>,
+) -> Result<HttpResponse, ApiError> {
+    new_account
+        .0
+        .validate()
+        .map_err(|err| ApiError::InvalidInput(validation_errors_to_string(err, None)))?;
+
+    if check_turnstile_captcha(&req, &new_account.challenge).await? {
+        return Err(ApiError::Turnstile);
+    }
+
+    if crate::database::models::User::get(&new_account.username, &**pool, &redis)
+        .await?
+        .is_some()
+    {
+        return Err(ApiError::InvalidInput("Username is taken!".to_string()));
+    }
+
+    let mut transaction = pool.begin().await?;
+    let user_id = crate::database::models::generate_user_id(&mut transaction).await?;
+
+    let new_account = new_account.0;
+
+    let score = zxcvbn::zxcvbn(
+        &new_account.password,
+        &[&new_account.username, &new_account.email],
+    )?;
+
+    if score.score() < 3 {
+        return Err(ApiError::InvalidInput(
+            if let Some(feedback) = score.feedback().clone().and_then(|x| x.warning()) {
+                format!("Password too weak: {}", feedback)
+            } else {
+                "Specified password is too weak! Please improve its strength.".to_string()
+            },
+        ));
+    }
+
+    let hasher = Argon2::default();
+    let salt = SaltString::generate(&mut ChaCha20Rng::from_entropy());
+    let password_hash = hasher
+        .hash_password(new_account.password.as_bytes(), &salt)?
+        .to_string();
+
+    crate::database::models::User {
+        id: user_id,
+        github_id: None,
+        discord_id: None,
+        gitlab_id: None,
+        google_id: None,
+        steam_id: None,
+        microsoft_id: None,
+        password: Some(password_hash),
+        username: new_account.username.clone(),
+        name: Some(new_account.username),
+        email: Some(new_account.email),
+        email_verified: false,
+        avatar_url: None,
+        bio: None,
+        created: Utc::now(),
+        role: Role::Developer.to_string(),
+        badges: Badges::default(),
+        balance: Decimal::ZERO,
+        payout_wallet: None,
+        payout_wallet_type: None,
+        payout_address: None,
+    }
+    .insert(&mut transaction)
+    .await?;
+
+    let session = issue_session(req, user_id, &mut transaction, &redis).await?;
+    let res = crate::models::sessions::Session::from(session, true);
+    transaction.commit().await?;
+
+    Ok(HttpResponse::Ok().json(res))
+}
+
+#[derive(Deserialize, Validate)]
+pub struct Login {
+    pub username: String,
+    pub password: String,
+    pub challenge: String,
+}
+
+#[post("login")]
+pub async fn login_password(
+    req: HttpRequest,
+    pool: Data<PgPool>,
+    redis: Data<deadpool_redis::Pool>,
+    login: web::Json<Login>,
+) -> Result<HttpResponse, ApiError> {
+    if check_turnstile_captcha(&req, &login.challenge).await? {
```

Comment:
Do we want this to be present on every login attempt? Turnstile is not as obnoxious as some other captchas but it is still indeed another step that some find annoying if not on the 3rd login attempt or smthn